### PR TITLE
LibWeb: Implement an ephemeral `URL.createObjectURL`

### PR DESCRIPTION
**Pull request by trflynn89**
Tested with the following HTML running on a local server:
```html
<html>
<body>
<script type="text/javascript">
    async function test() {
        const response = await fetch('./90s-bg.png');
        const data = await response.blob();
        return URL.createObjectURL(data);
    }

    test().then(url => {
        let image = new Image(240, 240)
        image.src = url;
        image.onload = () => {
            URL.revokeObjectURL(image.src);
        };

        document.body.appendChild(image);
    });
</script>
</body>
</html>
```

This test isn't added in-tree because CORS prevents the `fetch` invocation from working on `file://` URLs.

### CODE REVIEW

**Comment by trflynn89** [comment]
This is a draft for now because we are hitting a crash when streaming the blob response's body:
```
WebContent: Userland/Libraries/LibWeb/HTML/Scripting/Environments.cpp:350: Web::HTML::EnvironmentSettingsObject& Web::HTML::incumbent_settings_object(): Assertion `!event_loop.is_backup_incumbent_settings_object_stack_empty()' failed.
```

```
#7  0x00007ffff709efb2 in Web::HTML::incumbent_settings_object () at Userland/Libraries/LibWeb/HTML/Scripting/Environments.cpp:350
#8  0x00007ffff6ce112d in operator() () at Userland/Libraries/LibWeb/Bindings/MainThreadVM.cpp:317
#9  call () at Meta/Lagom/../../AK/Function.h:157
#10 0x00007ffff56d166f in AK::Function<JS::JobCallback (JS::FunctionObject&)>::operator()(JS::FunctionObject&) const () at Meta/Lagom/../../AK/Function.h:95
#11 0x00007ffff56cff44 in JS::Promise::perform_then () at Userland/Libraries/LibJS/Runtime/Promise.cpp:307
#12 0x00007ffff72b3153 in Web::WebIDL::react_to_promise(JS::PromiseCapability const&, AK::Optional<JS::SafeFunction<Web::WebIDL::ExceptionOr<JS::Value> (JS::Value)> >, AK::Optional<JS::SafeFunction<Web::WebIDL::ExceptionOr<JS::Value> (JS::Value)> >) () at Userland/Libraries/LibWeb/WebIDL/Promise.cpp:143
#13 0x00007ffff72b37eb in Web::WebIDL::upon_fulfillment(JS::PromiseCapability const&, JS::SafeFunction<Web::WebIDL::ExceptionOr<JS::Value> (JS::Value)>) () at Userland/Libraries/LibWeb/WebIDL/Promise.cpp:159
#14 0x00007ffff7201507 in Web::Streams::set_up_readable_byte_stream_controller(Web::Streams::ReadableStream&, Web::Streams::ReadableByteStreamController&, JS::SafeFunction<Web::WebIDL::ExceptionOr<JS::Value> ()>&&, JS::SafeFunction<Web::WebIDL::ExceptionOr<JS::NonnullGCPtr<JS::PromiseCapability> > ()>&&, JS::SafeFunction<Web::WebIDL::ExceptionOr<JS::NonnullGCPtr<JS::PromiseCapability> > (JS::Value)>&&, double, JS::Value) () at Userland/Libraries/LibWeb/Streams/AbstractOperations.cpp:1410
#15 0x00007ffff72051ed in Web::Streams::set_up_readable_stream_controller_with_byte_reading_support(Web::Streams::ReadableStream&, AK::Optional<JS::SafeFunction<Web::WebIDL::ExceptionOr<JS::NonnullGCPtr<JS::PromiseCapability> > ()> >&&, AK::Optional<JS::SafeFunction<Web::WebIDL::ExceptionOr<JS::NonnullGCPtr<JS::PromiseCapability> > (JS::Value)> >&&, double)
    () at Userland/Libraries/LibWeb/Streams/AbstractOperations.cpp:1736
#16 0x00007ffff6f538fb in Web::FileAPI::Blob::get_stream () at Userland/Libraries/LibWeb/FileAPI/Blob.cpp:274
#17 0x00007ffff6ef6061 in Web::Fetch::extract_body () at Userland/Libraries/LibWeb/Fetch/BodyInit.cpp:51
#18 0x00007ffff6ef8745 in Web::Fetch::safely_extract_body () at Userland/Libraries/LibWeb/Fetch/BodyInit.cpp:32
#19 0x00007ffff6f0a280 in Web::Fetch::Fetching::scheme_fetch () at Userland/Libraries/LibWeb/Fetch/Fetching/Fetching.cpp:744
#20 0x00007ffff6f0accb in operator() () at Userland/Libraries/LibWeb/Fetch/Fetching/Fetching.cpp:300
#21 0x00007ffff6f0f858 in call () at Meta/Lagom/../../Userland/Libraries/LibJS/SafeFunction.h:148
#22 JS::SafeFunction<Web::WebIDL::ExceptionOr<JS::NonnullGCPtr<Web::Fetch::Fetching::PendingResponse> > ()>::operator()() const () at Meta/Lagom/../../Userland/Libraries/LibJS/SafeFunction.h:81
#23 operator() () at Userland/Libraries/LibWeb/Fetch/Fetching/Fetching.cpp:385
```

I've no idea what to do here so any ideas welcome :)

File: Userland/Libraries/LibWeb/Fetch/Fetching/Fetching.cpp [hunk 1]
```diff
@@ -708,8 +710,83 @@ WebIDL::ExceptionOr<JS::NonnullGCPtr<PendingResponse>> scheme_fetch(JS::Realm& r
...
+        // 7. Let type be blob’s type.
+        auto const& type = blob->type();
+
+        // 8. If request’s header list does not contain `Range`:
+        if (!request->header_list()->contains("Range"sv.bytes())) {
+            // 1. Let bodyWithType be the result of safely extracting blob.
+            auto body_with_type = TRY(safely_extract_body(realm, blob));
+
+            // 2. Set response’s status message to `OK`.
+            response->set_status_message(TRY_OR_THROW_OOM(vm, ByteBuffer::copy("OK"sv.bytes())));
```
Comment (linusg) [inline Userland/Libraries/LibWeb/Fetch/Fetching/Fetching.cpp:747]: ByteBuffer has an inline capacity, this cannot OOM - see also:

image

File: Userland/Libraries/LibWeb/Fetch/Fetching/Fetching.cpp [hunk 2]
```diff
@@ -708,8 +710,83 @@ WebIDL::ExceptionOr<JS::NonnullGCPtr<PendingResponse>> scheme_fetch(JS::Realm& r
...
+            // 4. Set response’s header list to « (`Content-Length`, serializedFullLength), (`Content-Type`, type) ».
+            auto content_length_header = TRY_OR_THROW_OOM(vm, Infrastructure::Header::from_string_pair("Content-Length"sv, serialized_full_length));
+            TRY_OR_THROW_OOM(vm, response->header_list()->append(move(content_length_header)));
+
+            auto content_type_header = TRY_OR_THROW_OOM(vm, Infrastructure::Header::from_string_pair("Content-Type"sv, type));
+            TRY_OR_THROW_OOM(vm, response->header_list()->append(move(content_type_header)));
+        }
+        // 9. Otherwise:
+        else {
+            // FIXME: 1. Set response’s range-requested flag.
```
Comment (linusg) [inline Userland/Libraries/LibWeb/Fetch/Fetching/Fetching.cpp:761]: IMO we can just FIXME step 9 above but this also works :sweat_smile:

**Review by linusg — COMMENTED** [review]
Nice!

**Review by linusg — APPROVED** [review]

**Comment by shannonbooth** [comment]
Nice! I was just about to make a start on this, but you got there first! LGTM

**Pull request merged**